starlite/datastructures/headers.py

```diff
@@ -1,15 +1,396 @@
+    libs.org/en/stable/multidict.html#cimultidictproxy) for HTTP headers."""
+
+    def __init__(self, headers: Optional[Union[Mapping[str, str], "RawHeadersList"]] = None) -> None:
+        headers_: Union[Mapping[str, str], List[Tuple[str, str]]] = {}
+        if isinstance(headers, list):
+            headers_ = [(key.decode("latin-1"), value.decode("latin-1")) for key, value in headers]
+        elif headers:
+            headers_ = headers
+
+        super().__init__(CIMultiDict(headers_))
+
+    @classmethod
+    def from_scope(cls, scope: "Scope") -> "Headers":
+        """Create headers from a `Scope`.
+
+        Args:
+            scope: An ASGI Scope
+
+        Returns:
+            Headers
+        """
+        return cls(list(scope["headers"]))
+
+    @classmethod
+    def from_message(cls, message: "Message") -> "Headers":
+        """
+        Create headers from a send-message.
+        Args:
+            message: An message
+
+        Returns:
+            Headers
+
+        Raises:
+            ValueError: If the message does not have a `headers` key
+        """
+        try:
+            return cls(message["headers"])  # type: ignore[typeddict-item]
+        except KeyError as exc:
+            raise ValueError(f"Unsupported message type: {message['type']!r}") from exc
+
+    @property
+    def raw(self) -> "RawHeadersList":
+        """Raw header value.
+
+        Returns:
+            A list of tuples contain the header and header-value as bytes
+        """
+        return _encode_headers((key, value) for key in set(self) for value in self.getlist(key))
+
+    def keys(self) -> List[str]:  # type: ignore[override]
+        """Get a list of all header names. Contains duplicates.
+
+        Returns:
+            A list of strings
+        """
+        return list(super().keys())
+
+    def values(self) -> List[str]:  # type: ignore[override]
+        """Get a list of all header values, including values of duplicate
+        headers.
+
+        Returns:
+            A list of strings
+        """
+        return list(super().values())
+
+    def items(self) -> List[Tuple[str, str]]:  # type: ignore[override]
+        """Get a list of all headers and values.
+
+        Returns:
+            A list of tuples containing header names and values
+        """
+        return list(super().items())
+
+    @deprecated("1.36.0", alternative="Headers.getall")
+    def getlist(self, key: str) -> List[str]:
+        """Get all values of a header.
+
+        Args:
+            key: Name of the header
+
+        Returns:
+            A list of values
+        """
+        return super().getall(key, [])
+
+    def mutablecopy(self) -> "MutableHeaders":
+        """Create a mutable copy.
+
+        Returns:
+            A [MutableHeaders][starlite.datastructures.headers.MutableHeaders] instance
+        """
+        return MutableHeaders(self.raw[:])
+
+
+class MutableHeaders(Mapping):
```

Comment:
Why do we use only `CIMultiDictProxy`? The lib has mutable implementation as well.

I guess there is a good reason not to use it but I can not grasp it.

Reply:
Because we don't want a mutable multidict here. `Headers` should be immutable

Reply:
It doesn't make any sense. Word `Mutable` is literally in the class name and yet we want Headers to be immutable

Reply:
Oh, I get it now. I was confused before, sorry. This was done because the starlette version of this exposes internal datastrucutres that allow the headers to be mutated via side effects, which I tried to keep in an effort to be backwards compatible. The actual `CIMultiDictProxy` of course does not support this.